apps/aecore/src/aec_block_insertion.erl

```diff
@@ -228,25 +275,35 @@ build_insertion_ctx_prev(Node, PrevKeyNode) ->
                     %% Ok so the prev keyblock is present but not the prev block?
                     %% this shouldn't be the case even for the genesis block
                     {error, {illegal_orphan, node_hash(Node)}};
-                {ok, #node{type = key}} ->
-                    {error, prev_key_hash_inconsistency};
                 {ok, PrevNode} ->
-                    case node_prev_key_hash(PrevNode) =:= PrevKeyHash of
-                        true ->
+                    case {aec_chain_node:type(PrevNode), node_prev_key_hash(PrevNode) =:= PrevKeyHash} of
+                        {key, _} -> {error, prev_key_hash_inconsistency};
+                        {_, false} -> {error, prev_key_hash_inconsistency};
+                        {_, true} ->
                             %% Now assert heights
-                            build_insertion_ctx_check_prev_height(Node, PrevNode, PrevKeyNode);
-                        false ->
-                            {error, prev_key_hash_inconsistency}
+                            build_insertion_ctx_check_prev_height(Node, PrevNode, PrevKeyNode)
                     end
             end
    end.
 
-build_insertion_ctx_check_prev_height(#node{type = key} = Node, PrevNode, PrevKeyNode) ->
+-spec build_insertion_ctx_check_prev_height(
+    chain_node(), chain_node(), chain_node()) ->
+    {ok, chain_node(), chain_node()} | {error, atom()}.
+build_insertion_ctx_check_prev_height(Node, PrevNode, PrevKeyNode) ->
+    case aec_chain_node:type(Node) of
+        key ->
+            build_insertion_ctx_check_prev_height_key(Node, PrevNode, PrevKeyNode);
+        micro ->
+            build_insertion_ctx_check_prev_height_micro(Node, PrevNode, PrevKeyNode)
+    end.
```

Comment:
How about moving this `case` to the `build_insertion_ctx_check_prev_height` clause pattern-matching?
```erlang
build_insertion_ctx_check_prev_height(key, ...) -> ...
build_insertion_ctx_check_prev_height(micro, ...) -> ...
```